### PR TITLE
Enable Javalab to use Version History

### PR DESCRIPTION
**Pull request by hannahbergam**
Other project-backed labs, such as App Lab, Web Lab, Sprite Lab, and Game Lab, allow students to revert their projects to previous versions. This PR adds that same functionality to Java Lab.

That functionality already exists in the Code Studio ecosystem. This setup largely mimics that of Applab, but without utilizing the same CodeWorkspace and DataWorkspace, and instead bringing the version history button directly into the JavaLabEditor.  

The existing testing framework exists on VersionHistoryTest. VersionHistory is imported on StudioApp, which is then accessed from AppLap, JavaLab, etc. There are no existing tests for applab version history specifically, so this PR follows that example. The existing JavaLabView tests have been updated to include the new required versionHistory prop. 

https://codedotorg.atlassian.net/browse/CSA-248

image

image

### CODE REVIEW

**Review by jmkulwik — APPROVED** [review]
This looks great! One question: Does levelbuilder still work as expected? i.e. especially wondering about the edit_sources page.

File: apps/src/javalab/Javalab.js [hunk 1]
```diff
@@ -76,6 +76,10 @@ Javalab.prototype.init = function(config) {
   const onCommitCode = this.onCommitCode.bind(this);
   const onInputMessage = this.onInputMessage.bind(this);
 
+  Javalab.handleVersionHistory = this.studioApp_.getVersionHistoryHandler(
```
Comment (molly-moen) [inline apps/src/javalab/Javalab.js]: for my information: is there a reason this is `Javalab.handleVersionHistory` instead of `const handleVersionHistory`?

**Review by molly-moen — APPROVED** [review]
LGTM!

**Comment by hannahbergam** [comment]
> This looks great! One question: Does levelbuilder still work as expected? i.e. especially wondering about the edit_sources page.

LevelBuilder still looks good- can you be more specific? I'm not sure which section you mean for edit sources.

File: apps/src/javalab/Javalab.js (hunk 1 above)
Comment (hannahbergam) [reply apps/src/javalab/Javalab.js]: This is an excellent question... I noticed the distinction in AppLab and was trying to follow suit. That being said, I'm not sure I understand why! Happy to switch it and see if anything breaks.

**Comment by jmkulwik** [comment]
> > This looks great! One question: Does levelbuilder still work as expected? i.e. especially wondering about the edit_sources page.
> 
> LevelBuilder still looks good- can you be more specific? I'm not sure which section you mean for edit sources.

Each level on levelbuilder has a page `/edit_blocks/start_sources` that allows a levelbuilder to edit the code that exists in the level when the page loads. Starting over replaces any student code with the code that was set on the `start_sources` page. 

Example: The first allthethings Java Lab level is 26927 on my dev environment. I'd edit the start sources by... 
* going here http://localhost-studio.code.org:3000/levels/26927/edit_blocks/start_sources
* Changing the code that exists on the page
* Click the "Save" button at the top of the screen

Then if I go here: http://localhost-studio.code.org:3000/levels/26927 and click "start over" in version history, I should see the code I added to `start_sources`

File: apps/src/javalab/Javalab.js [hunk 2]
```diff
@@ -90,6 +90,15 @@ Javalab.prototype.init = function(config) {
     visualization = <DefaultVisualization />;
   }
 
+  Javalab.handleVersionHistory = this.studioApp_.getVersionHistoryHandler(
+    config
+  );
+
+  config.afterClearPuzzle = function() {
```
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: This should be defined outside the Javalab.init. Similar to how we handle getCode.

File: apps/src/javalab/Javalab.js (hunk 1 above)
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: +1 I'd also suggest moving it to const. Also moving it up near where we're setting the other handlers (i.e. by `onInputMessage` and `onCommitCode`)

File: apps/src/javalab/Javalab.js (hunk 2 above)
Comment (jmkulwik) [reply apps/src/javalab/Javalab.js]: (And setting it on the config should be moved up by `getCode`)

Comment (hannahbergam) [reply apps/src/javalab/Javalab.js]: Oh gotcha- I'll do this. Thanks!

File: apps/src/javalab/Javalab.js [hunk 3]
```diff
@@ -90,6 +91,8 @@ Javalab.prototype.init = function(config) {
     visualization = <DefaultVisualization />;
   }
 
+  const handleVersionHistory = this.studioApp_.getVersionHistoryHandler(config);
```
Comment (jmkulwik) [inline apps/src/javalab/Javalab.js]: Just a quick reminder to move this up to where the other handlers are declared (i.e. by `onInputMessage` above)

File: apps/src/javalab/JavalabView.jsx [hunk 4]
```diff
@@ -145,48 +150,20 @@ class JavalabView extends React.Component {
...
-    channelId: state.pageConstants.channelId,
-    isDarkMode: state.javalab.isDarkMode
-  }),
-  dispatch => ({
-    appendOutputLog: log => dispatch(appendOutputLog(log)),
-    setIsDarkMode: isDarkMode => dispatch(setIsDarkMode(isDarkMode))
-  })
-)(UnconnectedJavalabView);
-
 const styles = {
```
Comment (jmkulwik) [inline apps/src/javalab/JavalabView.jsx:153]: (Ignore if this is still a WIP)
It looks like some of these styles may have been unintentionally changed while resolving merge conflicts.
image
Which leads to an interesting view
image

File: apps/src/javalab/Javalab.js (hunk 1 above)
Comment (hannahbergam) [reply apps/src/javalab/Javalab.js]: Just saw this one- thanks!

File: apps/src/javalab/Javalab.js (hunk 3 above)
Comment (hannahbergam) [reply apps/src/javalab/Javalab.js]: 🙌

File: apps/src/javalab/JavalabView.jsx (hunk 4 above)
Comment (hannahbergam) [reply apps/src/javalab/JavalabView.jsx:153]: 🤦‍♀️Fixing now. Thank you

**Review by jmkulwik — APPROVED** [review]
✔️ for the new changes!

**Pull request merged**